cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java

```diff
@@ -0,0 +1,347 @@
+      runRecordCompletedHandler(appId, programId, programType);
+      programArgsHandler(appId, programId, programType);
+    }
+  }
+
+  /**
+   * Handles the {@link AppMetadataStore#TYPE_PROGRAM_ARGS} meta data and writes it back with namespace
+   *
+   * @param appId       the application id to which this program belongs to
+   * @param programId   the program id of the program
+   * @param programType the {@link ProgramType} of the program
+   */
+  private void programArgsHandler(final String appId, final String programId, final ProgramType programType) {
+    final byte[] partialKey = new MDSKey.Builder().add(AppMetadataStore.TYPE_RUN_RECORD_STARTED, DEVELOPER_ACCOUNT,
+                                                       appId, programId).build().getKey();
+    appMDS.executeUnchecked(new TransactionExecutor.Function<DatasetContext<Table>, Void>() {
+      @Override
+      public Void apply(DatasetContext<Table> input) throws Exception {
+        Scanner rows = input.get().scan(partialKey, Bytes.stopKeyForPrefix(partialKey));
+        Row row;
+        while ((row = rows.next()) != null) {
+          ProgramArgs programArgs = GSON.fromJson(Bytes.toString(row.get(COLUMN)), ProgramArgs.class);
+          store.storeRunArguments(Id.Program.from(Id.Application.from(Constants.DEFAULT_NAMESPACE, appId), programType,
+                                                  programId), programArgs.getArgs());
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Handles the {@link AppMetadataStore#TYPE_RUN_RECORD_STARTED} meta data and writes it back with namespace
+   *
+   * @param appId       the application id to which this program belongs to
+   * @param programId   the program id of the program
+   * @param programType the {@link ProgramType} of the program
+   */
+  private void runRecordStartedHandler(final String appId, final String programId, final ProgramType programType) {
+    final byte[] partialKey = new MDSKey.Builder().add(AppMetadataStore.TYPE_RUN_RECORD_STARTED, DEVELOPER_ACCOUNT,
+                                                       appId, programId).build().getKey();
+    appMDS.executeUnchecked(new TransactionExecutor.Function<DatasetContext<Table>, Void>() {
+      @Override
+      public Void apply(DatasetContext<Table> input) throws Exception {
+        Scanner rows = input.get().scan(partialKey, Bytes.stopKeyForPrefix(partialKey));
+        Row row;
+        while ((row = rows.next()) != null) {
+          RunRecord runRecord = GSON.fromJson(Bytes.toString(row.get(COLUMN)), RunRecord.class);
+
+          MDSKey.Splitter keyParts = new MDSKey(row.getRow()).split();
+          // skip runRecordStarted
+          keyParts.getString();
+          // skip default
+          keyParts.skipString();
+          // skip appId
+          keyParts.skipString();
+          // skip programId
+          keyParts.skipString();
+          String pId = keyParts.getString();
```

Comment:
Lets keep this pId here to be consistent with how the key was constructed else it will get too confusing.
If we want to rename it we will rename it both places in a different PR.